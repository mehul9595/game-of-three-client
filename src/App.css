.App {
  display: flex;
  flex-direction: column;
  align-items: center;
  /* background-color: #c5cedf; */
  background-color: #ffecd9;
  height: 100vh;
  width: 100%;
  margin: 0 auto;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}


.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
