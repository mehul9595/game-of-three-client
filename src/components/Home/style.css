.game-home {
    width: 100%;
    height: 100%;
}

.game-home .ant-btn.ant-btn-primary.game-button {
  height: 4em;
  font-size: 16px;
  width: 13em;
  margin: 10px;
  padding-left: 20px;
  border-radius: 50px;
  float: left;
  color: #fff;
  box-shadow: 5px 5px 2px -2px #a5a5a5;
}
