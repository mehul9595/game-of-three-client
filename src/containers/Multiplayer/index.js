import React, { useState, useEffect } from "react";
import Header from "../../components/Header";
import PlayArea from "../../components/PlayArea";
import ActionButton from "../../components/ActionButton";
import useMultiplayer from "../Hooks/useMultiplayer";
import { message } from "antd";

const Multiplayer = (props) => {
  const [isButtonDisabled, setIsButtonDisabled] = useState(false);
  const { playerData, position, calculateUserAction } = useMultiplayer();

  useEffect(() => {
    console.log("playerData:@18", playerData);
    if (playerData && playerData.turnArray) {
      var lastTurn = playerData.turnArray[playerData.turnCount];

      if (lastTurn.value === 1) {
        setIsButtonDisabled(true);
        //this.modalForGameResult(lastTurn.player);
        message.success(`${lastTurn.player} Wins`);
      } else {
        //   if (playerData.turnCount !== 0) {
        lastTurn.player === position
          ? setIsButtonDisabled(true)
          : setIsButtonDisabled(false);
        //   }
      }
    }
  }, [playerData, position]);

  //position, playerData
  const actionButtonHandler = (params) => {
    console.log(params);

    let currentTurnValue = playerData.turnArray[playerData.turnCount].value;
    if ((currentTurnValue + params) % 3 === 0) {
      calculateUserAction({
        requiredNumberToBeDividedByThree: params,
        turnCount: playerData.turnCount + 1,
      });
    } else {
      //   notify(`${currentTurnValue} can not be divided by ${actionValue}`);
      alert(`${currentTurnValue} can not be divided by ${params}`);
    }
  };

  return (
    <div>
      <Header />
      {!playerData.isGameStart ? (
        <h3>Waiting for another player...</h3>
      ) : (
        <>
          <PlayArea
            turnArray={playerData.turnArray}
            playerTurn={playerData.playerTurn}
            gameMode={"multiplayer"}
          />
          <ActionButton
            actionHandler={actionButtonHandler}
            isDisabled={isButtonDisabled}
          />
        </>
      )}
    </div>
  );
};

export default Multiplayer;
