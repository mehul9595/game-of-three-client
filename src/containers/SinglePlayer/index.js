import React, { useState, useEffect } from "react";
import Header from "../../components/Header";
import PlayArea from "../../components/PlayArea";
import ActionButton from "../../components/ActionButton";
import { message } from "antd";
import "antd/dist/antd.css";

const getInitialNumber = () => {
  let randomNumber = parseInt(Math.random() * 100, 10);
  if (randomNumber < 2) getInitialNumber();
  else return randomNumber;
};

const playerTypes = {
  bot: "bot",
  player: "player",
};

const SinglePlayer = (props) => {
  const [turnCount, setTurnCount] = useState(0);
  // const [
  //   requiredNumberToBeDividedByThree,
  //   setRequiredNumberToBeDividedByThree,
  // ] = useState(null);
  let requiredNumberToBeDividedByThree = null;
  const [turnArray, setTurnArray] = useState([
    {
      id: 0,
      player: playerTypes.bot,
      value: getInitialNumber(),
      action: undefined,
      valueExpression: "",
    },
  ]);

  useEffect(() => {
    // if (turnCount > 0 && turnCount % 2 === 0) {
    //   let botTurn = turnCount + 1;
    //   calculateNumbers(botTurn);
    //   console.log("turnCount", botTurn);
    //   setTurnCount(botTurn);
    //   if (calculateNewNumberToSendOpponent(botTurn).value === 1) {
    //     // this.modalForGameResult(this.detectTurnPlayer());
    //     message.success("win");
    //     return true;
    //   }
    // } else {
      calculateNumbers(turnCount);
      console.log("turnCount", turnCount);
      if (calculateNewNumberToSendOpponent(turnCount).value === 1) {
        // this.modalForGameResult(this.detectTurnPlayer());
        message.success("win");
        return true;
      }
    // }
  }, [turnCount]);

  const assignNumberNeedsToBeAdded = () => {
    var turnValue = turnArray[turnCount - 1].value;
    console.log(turnValue);
    console.log("turnCOunt", turnCount);
    if (turnValue % 3 === 0) {
      requiredNumberToBeDividedByThree = 0;
    } else if ((turnValue + 1) % 3 === 0) {
      requiredNumberToBeDividedByThree = 1;
    } else if ((turnValue - 1) % 3 === 0) {
      requiredNumberToBeDividedByThree = -1;
    }
  };

  const updateTurnArray = (newTurn) => {
    setTurnArray((result) => [...result, newTurn]);
  };

  const getPlayerTurn = (params) => {
    let turnPlayer = params % 2 === 0 ? playerTypes.bot : playerTypes.player;
    return turnPlayer;
  };

  const setNewTurn = (turnCounter) => {
    var calcNewNumber = calculateNewNumberToSendOpponent(turnCounter);
    let newTurn = {
      id: turnCounter,
      player: getPlayerTurn(turnCounter),
      value: calcNewNumber.value,
      action: requiredNumberToBeDividedByThree,
      valueExpression: calcNewNumber.valueExpresionStr,
    };

    updateTurnArray(newTurn);
  };

  const calculateNewNumberToSendOpponent = (turnCounter) => {
    let expression =
      turnArray[turnCounter - 1].value + requiredNumberToBeDividedByThree;

    let calculatedValue = parseInt(expression, 10) / 3;

    let stringExpression = `[( ${requiredNumberToBeDividedByThree} + ${
      turnArray[turnCounter - 1].value
    } ) / 3] = ${calculatedValue}`;
    console.log("stringExpression: ", stringExpression);

    return {
      value: calculatedValue,
      valueExpresionStr: stringExpression,
    };
  };

  const calculateNumbers = (turnCount) => {
    let turnValue = turnArray[turnCount - 1].value;

    assignNumberNeedsToBeAdded(turnValue);
    setNewTurn(turnCount);
  };

  const checkActionAndResult = (action) => {
    message.info(action);
    let actionValue = parseInt(action);
    let currentValue = turnArray[turnCount].value;

    if ((currentValue + actionValue) % 3 === 0) {
      message.success("OK good");
      let turnCounter = turnCount + 1;
      setTurnCount(turnCounter);

      // after player turns, set turn count. on setTimeout 1 sec call checkAndResult for bot. action value would be previous turn count value
      // setTimeout(async ()=> {

      //   console.log("new turn cout after 1 sec", turnCounter);
      //   console.log("new array after 1 sec ", turnArray);

      //   // assignNumberNeedsToBeAdded(turnCount+1);

      // }, 5000);
    } else {
      message.error("Number cannot be divided by 3");
    }
  };

  return (
    <>
      <Header />
      <PlayArea turnArray={turnArray} />
      <ActionButton actionHandler={checkActionAndResult} />
    </>
  );
};

export default SinglePlayer;
