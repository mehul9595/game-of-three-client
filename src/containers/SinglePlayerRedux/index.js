import React, { useEffect } from "react";
import Header from "../../components/Header";
import PlayArea from "../../components/PlayArea";
import ActionButton from "../../components/ActionButton";
import ModalResult from "../../components/Modal";
import Helper from "../../utils/helper";
import { message } from "antd";
import "antd/dist/antd.css";
import {
  PLAYER_ACTION,
  NUMBER_TO_DIVIDE_BY_THREE,
  TURN_COUNTER,
  BOT_ACTION,
  INITIALIZE_GAME_PLAY,
} from "../../redux/actionTypes";
import { useDispatch, useSelector } from "react-redux";

const SinglePlayerRedux = (props) => {
  const playerData = useSelector((state) => state.PlayerData);
  const dispatch = useDispatch();

  useEffect(() => {
    if (playerData.winner === true) {
      console.log(playerData.turnCount);

      let lastTurn = playerData.turnArray[playerData.turnCount];
      // console.log("last turn under winner effect", lastTurn);
      
      setTimeout(() => {
        ModalResult("", lastTurn.player === Helper.playerTypes.player, () => {
          props.history.push("/");
          dispatch({ type: INITIALIZE_GAME_PLAY });
        });  
      }, 500);

      return true;
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [playerData, props]);

  const onActionButtonClick = (action) => {
    let actionValue = parseInt(action);
    let currentValue = playerData.turnArray[playerData.turnCount].value;

    if ((currentValue + actionValue) % 3 === 0) {
      dispatch({ type: TURN_COUNTER });

      dispatch({ type: NUMBER_TO_DIVIDE_BY_THREE });

      dispatch({
        type: PLAYER_ACTION,
        payload: action,
      });

      dispatch({
        type: BOT_ACTION,
        payload: {},
      });
    } else {
      message.error(`${currentValue} can not be divided by ${action}`);
    }
  };

  return (
    <>
      <Header />
      <PlayArea turnArray={playerData.turnArray} />
      <ActionButton actionHandler={onActionButtonClick} />
    </>
  );
};

export default SinglePlayerRedux;
