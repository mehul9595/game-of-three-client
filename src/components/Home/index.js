import React from "react";
import { Card, Button, Divider, Row, Col } from "antd";
import { PlayCircleOutlined } from "@ant-design/icons";
import "./style.css";
import "antd/dist/antd.css";

const Home = (props) => {
  return (
    <Card bordered hoverable className="game-home">
      <h1>Game of Three @ Scoober</h1>
      <Divider className="game-home-divider" />
      <Row justify="center">
        <Col>
          <Button
            type="primary"
            icon={<PlayCircleOutlined />}
            htmlType="button"
            className="game-button"
            onClick={() => props.history.push("/singleplayer")}
          >
            Single Player
          </Button>
        </Col>
      </Row>
      <Row justify="center">
        <Col>
          <Button
            type="primary"
            htmlType="button"
            className="game-button"
            icon={<PlayCircleOutlined />}
            onClick={() => props.history.push("/multiplayer")}
            data-testId="multiplayerBtn"
          > 
            Multiplayer
          </Button>
        </Col>
      </Row>
    </Card>
  );
};

export default Home;
