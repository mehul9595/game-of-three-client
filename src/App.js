import React from "react";
import Header from "./components/Header";
import Layout from "./containers/Layout";
import Home from './components/Home'
import PlayArea from './components/PlayArea';
import ActionButton from "./components/ActionButton";
import "./App.css";

function App() {
  return (
    <div className="App">
      <Layout>
        <Header />
        <Home />
        <PlayArea />
        <ActionButton  />
      </Layout>
    </div>
  );
}

export default App;
