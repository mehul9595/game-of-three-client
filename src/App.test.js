import App from "./App";
import { shallow } from "enzyme";
import { Route } from "react-router-dom";
import Home from "./components/Home";
import SinglePlayer from './containers/SinglePlayer';
import Multiplayer from './containers/Multiplayer';
import SinglePlayerRedux from './containers/SinglePlayerRedux';

let pathMap = {};
describe("AppComponent", () => {
  const wrapper = shallow(<App />);
  
  beforeAll(() => {
    pathMap = wrapper.find(Route).reduce((pathMap, route) => {
      const routeProps = route.props();
      pathMap[routeProps.path] = routeProps.component;
      return pathMap;
    }, {});
  });

  test("should contain 4 routes", () => {
    
    expect(wrapper.find(Route)).toHaveLength(5);
    
  });

  test("should show MainScreen component", () => {
    expect(pathMap["/"]).toBe(Home);
  });

  test("should show SinglePlayer component", () => {
    expect(pathMap["/singleplayer"]).toBe(SinglePlayer);
  });

  test("should show MultiPlayer component", () => {
    expect(pathMap["/multiplayer"]).toBe(Multiplayer);
  });

  test("should show Single player redux component", () => {
    expect(pathMap["/singleplayer-redux"]).toBe(SinglePlayerRedux);
  });

  test("should show page not found case", () => {
    expect(pathMap[undefined]()).toMatchObject(<div>Page not found...</div>);
  });
});
