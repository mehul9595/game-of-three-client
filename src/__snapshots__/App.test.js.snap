// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with snapshot correctly 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="container"
    >
      <header
        class="App-header"
      >
        <img
          alt="logo"
          class="App-logo"
          src="logo.svg"
        />
        <div
          class="title"
        >
          <div
            class="first-line"
          >
            Scoober - Takeaway.com
          </div>
          <div
            class="sub-title"
          >
            Win the game or win the job
          </div>
        </div>
      </header>
      <div
        class="home"
      >
        <h1>
          Game of ThreeX
        </h1>
        <p>
          Single Player
        </p>
        <p>
          Multiplayer
        </p>
      </div>
      <div
        class="play-area-container"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
      </div>
      <div
        class="action-button-container"
      >
        <button
          class="ant-btn ant-btn-primary ant-btn-circle ant-btn-lg"
          type="button"
        >
          <span>
            +1
          </span>
        </button>
        <button
          class="ant-btn ant-btn-primary ant-btn-circle ant-btn-lg"
          type="button"
        >
          <span>
            0
          </span>
        </button>
        <button
          class="ant-btn ant-btn-primary ant-btn-circle ant-btn-lg"
          type="button"
        >
          <span>
            -1
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
